File: backend/app/service/notification.py
from typing import List, Optional
from fastapi import HTTPException
from sqlalchemy.orm import Session
from sqlalchemy import desc

from app.model.notification import Notification
from app.schema.notification import NotificationUpdate


def get_notification(db: Session, notification_id: int) -> Notification:
    """Get a notification by ID."""
    notification = db.query(Notification).filter(Notification.id == notification_id).first()
    if not notification:
        raise HTTPException(status_code=404, detail="Notification not found")
    return notification


def get_user_notifications(db: Session, user_id: int, skip: int = 0, limit: int = 20) -> List[Notification]:
    """Get all notifications for a user."""
    return db.query(Notification).filter(
        Notification.user_id == user_id
    ).order_by(desc(Notification.created_at)).offset(skip).limit(limit).all()


def count_user_notifications(db: Session, user_id: int) -> int:
    """Count the total number of notifications for a user."""
    return db.query(Notification).filter(Notification.user_id == user_id).count()


def count_unread_notifications(db: Session, user_id: int) -> int:
    """Count the number of unread notifications for a user."""
    return db.query(Notification).filter(
        (Notification.user_id == user_id) &
        (Notification.is_read == False)
    ).count()


def mark_notification_as_read(db: Session, notification_id: int, user_id: int) -> Notification:
    """Mark a notification as read."""
    db_notification = get_notification(db, notification_id)

    # Check if notification belongs to user
    if db_notification.user_id != user_id:
        raise HTTPException(status_code=403, detail="Not authorized to update this notification")

    # Update notification
    db_notification.is_read = True
    db.commit()
    db.refresh(db_notification)

    return db_notification


def mark_all_notifications_as_read(db: Session, user_id: int) -> int:
    """Mark all notifications for a user as read."""
    # Update all unread notifications
    result = db.query(Notification).filter(
        (Notification.user_id == user_id) &
        (Notification.is_read == False)
    ).update({"is_read": True})

    db.commit()

    return result


def delete_notification(db: Session, notification_id: int, user_id: int) -> bool:
    """Delete a notification."""
    db_notification = get_notification(db, notification_id)

    # Check if notification belongs to user
    if db_notification.user_id != user_id:
        raise HTTPException(status_code=403, detail="Not authorized to delete this notification")

    db.delete(db_notification)
    db.commit()

    return True


def delete_all_notifications(db: Session, user_id: int) -> int:
    """Delete all notifications for a user."""
    # Delete all notifications
    result = db.query(Notification).filter(Notification.user_id == user_id).delete()

    db.commit()

    return result

File: frontend/src/services/notification.js
import { db } from '../firebase';
import { collection, query, where, orderBy, getDocs, updateDoc, doc, serverTimestamp } from 'firebase/firestore';
import api from './api';

// Get all notifications
export const getNotifications = async (userId) => {
  const q = query(
    collection(db, 'notifications'),
    where('userId', '==', userId),
    orderBy('createdAt', 'desc')
  );
  const snapshot = await getDocs(q);
  return snapshot.docs.map(doc => ({ id: doc.id, ...doc.data() }));
};

// Get unread notifications count
export const getUnreadCount = async (userId) => {
  const q = query(
    collection(db, 'notifications'),
    where('userId', '==', userId),
    where('isRead', '==', false)
  );
  const snapshot = await getDocs(q);
  return snapshot.size;
};

// Mark a notification as read
export const markAsRead = async (notificationId) => {
  const notificationRef = doc(db, 'notifications', notificationId);
  await updateDoc(notificationRef, {
    isRead: true,
    readAt: serverTimestamp()
  });
};

// Mark all notifications as read
export const markAllAsRead = async (userId) => {
  const q = query(
    collection(db, 'notifications'),
    where('userId', '==', userId),
    where('isRead', '==', false)
  );
  const snapshot = await getDocs(q);
  
  const batch = db.batch();
  snapshot.docs.forEach(doc => {
    batch.update(doc.ref, {
      isRead: true,
      readAt: serverTimestamp()
    });
  });
  
  await batch.commit();
};

// Delete a notification
export const deleteNotification = async (notificationId) => {
  const response = await api.delete(`/notifications/${notificationId}`);
  return response.data;
};

// Delete all notifications
export const deleteAllNotifications = async () => {
  const response = await api.delete('/notifications');
  return response.data;
};